File: Android App/ColonelFund/app/src/main/java/com/colonelfund/colonelfund/Event.java
package com.colonelfund.colonelfund;

import org.json.JSONObject;
import org.json.JSONException;
import java.io.Serializable;

/**
 * Basic Event Object that holds all of the information for an event.
 */
public class Event implements Serializable {
    private static final long serialVersionUID = -8645384342222081215L;
    private String title;
    private String associatedMember;
    private String associatedEmail;
    private String eventDate;
    private double fundGoal;
    private double currentFunds;
    private String description;
    private String type;
    private String imageURL;
    private String eventTime;
    private Address address;

    /**
     * Event object constructor.
     *
     * @param title of an event.
     * @param associatedMember of an event.
     * @param eventDate of an event.
     * @param fundGoal of an event.
     * @param currentFunds of an event.
     * @param description of an event.
     * @param type of an event.
     */
    public Event(String title,
                 String associatedMember,
                 String eventDate,
                 double fundGoal,
                 double currentFunds,
                 String description,
                 String type) {
        super();
        this.title = title;
        this.associatedMember = associatedMember;
        this.eventDate = eventDate;
        this.fundGoal = fundGoal;
        this.currentFunds = currentFunds;
        this.description = description;
        this.type = type;
    }

    /**
     * Full Event object constructor.
     *
     * @param title of an event.
     * @param associatedMember of an event.
     * @param associatedEmail of an event.
     * @param eventDate of an event.
     * @param fundGoal of an event.
     * @param currentFunds of an event.
     * @param description of an event.
     * @param type of an event.
     * @param imageURL of an event.
     * @param address of an event.
     */
    public Event(String title,
                 String associatedMember,
                 String associatedEmail,
                 String eventDate,
                 double fundGoal,
                 double currentFunds,
                 String description,
                 String type,
                 String imageURL,
                 String eventTime,
                 Address address) {
        super();
        this.title = title;
        this.associatedMember = associatedMember;
        this.associatedEmail = associatedEmail;
        this.eventDate = eventDate;
        this.fundGoal = fundGoal;
        this.currentFunds = currentFunds;
        this.description = description;
        this.type = type;
        this.imageURL = imageURL;
        this.eventTime = eventTime;
        this.address = address;
    }

    /**
     * @return the title
     */
    public String getTitle() {
        return title;
    }

    /**
     * @param title the title to set
     */
    public void setTitle(String title) {
        this.title = title;
    }

    /**
     * @return the associatedMember
     */
    public String getAssociatedMember() {
        return associatedMember;
    }

    /**
     * @param associatedMember the associatedMember to set
     */
    public void setAssociatedMember(String associatedMember) {
        this.associatedMember = associatedMember;
    }

    /**
     * @param associatedEmail of an event. .
     */
    public void setAssociatedEmail(String associatedEmail) {
        this.associatedEmail = associatedEmail;
    }

    /**
     * @return the eventDate
     */
    public String getEventDate() {
        return eventDate;
    }

    /**
     * @param eventDate the eventDate to set
     */
    public void setEventDate(String eventDate) {
        this.eventDate = eventDate;
    }

    /**
     * @return the fundGoal
     */
    public double getFundGoal() {
        return fundGoal;
    }

    /**
     * @param fundGoal the fundGoal to set
     */
    public void setFundGoal(double fundGoal) {
        this.fundGoal = fundGoal;
    }

    /**
     * @return the currentFunds
     */
    public String getAssociatedEmail() {
        return associatedEmail;
    }

    /**
     * @return current funds of an event.
     */
    public double getCurrentFunds() {
        return currentFunds;
    }

    /**
     * @param currentFunds the currentFunds to set
     */
    public void setCurrentFunds(double currentFunds) {
        this.currentFunds = currentFunds;
    }

    /**
     * @return the description
     */
    public String getDescription() {
        return description;
    }

    /**
     * @param description
     */
    public void setDescription(String description) {
        this.description = description;
    }

    /**
     * @return the title
     */
    public String getType() {
        return type;
    }

    /**
     * @param type the title to set
     */
    public void setType(String type) {
        this.type = type;
    }

    /**
     * Get a JSONObject representation of Event
     *
     * @return JSONObject of an event
     */
    public String getImageURL() {
        return imageURL;
    }

    /**
     * @param imageURL of an event to set
     */
    public void setImageURL(String imageURL) {
        this.imageURL = imageURL;
    }

    /**
     *
     * @return the time of the event
     */
    public String getEventTime() {
        return eventTime;
    }

    /**
     *
     * @param eventTime of the event to set
     */
    public void setEventTime(String eventTime) {
        this.eventTime = eventTime;
    }

    /**
     *
     * @return address of event location
     */
    public Address getAddress() {
        return address;
    }

    /**
     *
     * @param address of event location to set
     */
    public void setAddress(Address address) {
        this.address = address;
    }

    /**
     * @return an event json.
     * @throws JSONException for event json creation.
     */
    public JSONObject toJson() throws JSONException {
        JSONObject jsonObj = new JSONObject();
        jsonObj.put("title", title);
        jsonObj.put("associatedMember", associatedMember);
        jsonObj.put("associatedEmail", associatedEmail);
        jsonObj.put("eventDate", eventDate);
        jsonObj.put("fundGoal", fundGoal);
        jsonObj.put("currentFunds", currentFunds);
        jsonObj.put("description", description);
        jsonObj.put("type", type);
        jsonObj.put("imageURL", imageURL);
        jsonObj.put("eventTime", eventTime);
        jsonObj.put("address",address.toJson());
        return jsonObj;
    }

    /**
     * Constructor with JSON Object.
     *
     * @param jsonObject of an event
     */
    public Event(JSONObject jsonObject) throws JSONException {
        this.title = jsonObject.getString("title");
        this.associatedMember = jsonObject.getString("associatedMember");
        this.associatedEmail = jsonObject.getString("associatedEmail");
        this.eventDate = jsonObject.getString("eventDate");
        this.fundGoal = jsonObject.getDouble("fundGoal");
        this.currentFunds = jsonObject.getDouble("currentFunds");
        this.description = jsonObject.getString("description");
        this.type = jsonObject.getString("type");
        this.imageURL = jsonObject.getString("imageURL");
        this.eventTime = jsonObject.getString("eventTime");
        this.address = new Address(jsonObject.getJSONObject("address"));
    }
}

File: iOS App/ColonelFund/ColonelFund/Event.swift
//
//  Event.swift
//  ColonelFund
//
//  Created by Chris Penner on 1/30/18.
//  Copyright © 2018 PennerTech. All rights reserved.
//

import Foundation

class Event: NSObject, Codable {
    
    private var title: String
    private var eventDate: String
    private var eventDescription: String
    private var fundGoal: Double
    private var currentFunds: Double
    private var eventPicURL: String
    private var eventPicData: String
    private var associatedMember: String
    private var eventType: String
    private var eventTime: String
    private var address: Address
    private var associatedEmail: String //delete
    
    init(title: String) {
        self.title = title
        self.eventDate = ""
        self.eventDescription = ""
        self.fundGoal = 0
        self.currentFunds = 0
        self.eventPicURL = ""
        self.eventPicData = "default"
        self.associatedMember = ""
        self.eventType = ""
        self.eventTime = ""
        self.address = Address()
        self.associatedEmail = "" //delete
    }
    
    init(title: String, eventDate: String, eventDescription: String, fundGoal: Double, currentFunds: Double, associatedMember: String, eventType: String) {
        self.title = title
        self.eventDate = eventDate
        self.eventDescription = eventDescription
        self.fundGoal = fundGoal
        self.currentFunds = currentFunds
        self.eventPicURL = ""
        self.eventPicData = "default"
        self.associatedMember = associatedMember
        self.eventType = eventType
        self.eventTime = ""
        self.address = Address()
        self.associatedEmail = "" //delete
    }
    
    init(title: String, eventDate: String, eventDescription: String, fundGoal: Double, currentFunds: Double, eventPicURL: String, associatedMember: String, eventType: String) {
        self.title = title
        self.eventDate = eventDate
        self.eventDescription = eventDescription
        self.fundGoal = fundGoal
        self.currentFunds = currentFunds
        self.eventPicURL = eventPicURL
        self.eventPicData = "default"
        self.associatedMember = associatedMember
        self.eventType = eventType
        self.eventTime = ""
        self.address = Address()
        self.associatedEmail = "" //delete
    }
    
    init(title: String, eventDate: String, eventDescription: String, fundGoal: Double, currentFunds: Double, eventPicData: String, associatedMember: String, eventType: String, eventTime: String, address: Address) {
        self.title = title
        self.eventDate = eventDate
        self.eventDescription = eventDescription
        self.fundGoal = fundGoal
        self.currentFunds = currentFunds
        self.eventPicURL = ""
        self.eventPicData = eventPicData
        self.associatedMember = associatedMember
        self.eventType = eventType
        self.eventTime = eventTime
        self.address = address
        self.associatedEmail = "" //delete
    }
    
    required init(from decoder: Decoder) throws {
        let values = try decoder.container(keyedBy: CodingKeys.self)
        title = try values.decode(String.self, forKey: .title)
        eventDate = try values.decode(String.self, forKey: .eventDate)
        eventDescription = try values.decode(String.self, forKey: .eventDescription)
        fundGoal = (try Double(values.decode(String.self, forKey: .fundGoal)))!
        currentFunds = (try Double(values.decode(String.self, forKey: .currentFunds)))!
        eventPicURL = try values.decode(String.self, forKey: .eventPicURL)
        associatedMember = try values.decode(String.self, forKey: .associatedMember)
        eventType = try values.decode(String.self, forKey: .eventType)
        associatedEmail = try values.decode(String.self, forKey: .associatedEmail) //delete
        self.eventPicData = "default"
        //TODO: fix below once database updated with eventTime and address
        self.eventTime = ""
        self.address = Address()
    }
    
    func getTitle() -> String {
        return self.title
    }
    
    func setTitle(title: String) {
        self.title = title
    }
    
    func getEventDate() -> String {
        return self.eventDate
    }
    
    func setEventDate(eventDate: String) {
        self.eventDate = eventDate
    }
    
    func getEventDescription() -> String {
        return self.eventDescription
    }
    
    func setEventDescription(eventDescription: String) {
        self.eventDescription = eventDescription
    }
    
    func getFundGoal() -> Double {
        return self.fundGoal
    }
    
    func setFundGoal(fundGoal: Double) {
        self.fundGoal = fundGoal
    }
    
    func getCurrentFunds() -> Double {
        return self.currentFunds
    }
    
    func setCurrentFunds(currentFunds: Double) {
        self.currentFunds = currentFunds
    }
    
    func getEventPicURL() -> String {
        return self.eventPicURL
    }
    
    func setEventPicURL(eventPicURL: String) {
        self.eventPicURL = eventPicURL
    }
    
    func getAssociatedMember() -> String {
        return associatedMember
    }
    
    func setAssociatedMember(associatedMember: String) {
        self.associatedMember = associatedMember
    }
    
    func getEventType() -> String {
        return eventType
    }
    
    func setEventType(eventType: String) {
        self.eventType = eventType
    }
    
    func getEventTime() -> String {
        return eventTime
    }
    
    func setEventTime(eventTime: String) {
        self.eventTime = eventTime
    }
    
    func getAddress() -> Address {
        return address
    }
    
    func setAddress(address: Address) {
        self.address = address
    }
    
    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(title, forKey: .title)
        try container.encode(eventDate, forKey: .eventDate)
        try container.encode(eventDescription, forKey: .eventDescription)
        try container.encode(String(fundGoal), forKey: .fundGoal)
        try container.encode(String(currentFunds), forKey: .currentFunds)
        try container.encode(eventPicData, forKey: .eventPicData)
        try container.encode(associatedMember, forKey: .associatedMember)
        try container.encode(eventType, forKey: .eventType)
        //TODO: uncomment below once database updated with eventTime and address
//        try container.encode(eventTime, forKey: .eventTime)
//        try container.encode(address, forKey: .address)
        //try container.encode(associatedEmail, forKey: .associatedEmail) //delete
    }
    
    func toFormEncoded() -> String {
        return "title=\(self.title)&associatedMember=\(self.associatedMember)&eventDate=\(self.eventDate)&fundGoal=\(self.fundGoal)&currentFunds=\(self.currentFunds)&description=\(self.eventDescription)&type=\(self.eventType)&image=\(self.eventPicData)"
        //TODO: uncomment below once database updated with eventTime and address
//        return "title=\(self.title)&associatedMember=\(self.associatedMember)&eventDate=\(self.eventDate)&fundGoal=\(self.fundGoal)&currentFunds=\(self.currentFunds)&description=\(self.eventDescription)&type=\(self.eventType)&image=\(self.eventPicData)&eventTime=\(self.eventTime)&address=\(self.address.toString())"
    }
    
    enum CodingKeys: String, CodingKey {
        case title
        case eventDate
        case eventDescription = "description"
        case fundGoal
        case currentFunds
        case eventPicURL = "imageURL"
        case eventPicData = "image"
        case associatedMember
        case eventType = "type"
        case eventTime
        case address
        case associatedEmail //delete
    }
}
